### PR TITLE
implement a multistream transport protocol

### PR DESCRIPTION
**Pull request by whyrusleeping**
@jbenet The peerstream tests for this fail. but i'm really not sure why. Everything passes fine when i integrate it into go-ipfs... (which doesnt mean much, but still).

Maybe you could take a look at this for me?

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -166,9 +166,15 @@ func (s *Swarm) addConn(netConn net.Conn, isServer bool) (*Conn, error) {
 	// unlocked **before** the Handler is called (which needs to be
 	// sequential). This was the simplest thing :)
 	setupConn := func() (*Conn, error) {
+
+		// create a new spdystream connection
+		ssConn, err := s.transport.NewConn(netConn, isServer)
+		if err != nil {
+			return nil, err
+		}
+
```
Comment (jbenet) [inline conn.go]: why is this getting moved up?? i dont think it should. see the note below:

```
// first, check if we already have it...
```

`transport.NewConn` could be an expensive operation.

**Comment by jbenet** [comment]
beautiful PR. :) :+1: :heart: composability.

File: conn.go (hunk 1 above)
Comment (whyrusleeping) [inline conn.go]: its a locking issue. if a swarm tries to connect to itself it will hang here.

Comment (jbenet) [inline conn.go]: oh interesting. setting up a connection _could_ be expensive (like RTT expensive). what about something like: a32d27b276851563ddfca718c012d1049e3155ae

Comment (whyrusleeping) [inline conn.go]: if doing things that way, we should put a placeholder connection in the map until we finish the handshake, that way we dont get any weird race conditions.

**Comment by whyrusleeping** [comment]
fixed a small bug.

**Comment by whyrusleeping** [comment]
the big test still fails though, And i'm not sure why...

**Comment by jbenet** [comment]
@whyrusleeping i force-pushed

**Comment by jbenet** [comment]
@whyrusleeping can you run this in some beefy machine? running the tests without `-race` all pass on my machine. with `-race` it hoses my laptop and travis

**Comment by jbenet** [comment]
LGTM

**Pull request merged**